engineering-decision-records/010-ledger-validation-context.md

```diff
@@ -0,0 +1,321 @@
+
+We aim to fuse those steps by allowing the slices to collect _local validation constraints_ while traversing the transaction. Hence, when validating a certificate, one can require witness validation later without processing the witnesses immediately. This minimizes the number of traversals we must perform on the transaction fields and lowers the cost to an `O(n)` in the case of witnesses. For now, we introduce a `WitnessSlice` to capture this behaviour.
+
+```rust
+pub trait WitnessSlice {
+    /// Indicate that a witness is required to be present (and valid) for the corresponding
+    /// set of credentials.
+    fn require_witness(&mut self, credential: StakeCredential);
+
+    /// Indicate that a bootstrap witness is required to be present (and valid) for the corresponding
+    /// root.
+    fn require_bootstrap_witness(&mut self, root: Hash<28>);
+
+    /// Obtain the full list of required signers collected while traversing the transaction.
+    fn required_signers(&mut self) -> BTreeSet<Hash<28>>;
+
+    /// Obtain the full list of required bootstrap witnesses collected while traversing the
+    /// transaction.
+    fn required_bootstrap_signers(&mut self) -> BTreeSet<Hash<28>>;
+}
+```
+
+### Further considerations
+
+The current design will effectively process blocks in three passes:
+
+1. Parsing
+2. Preparation
+3. Validation
+
+It sounds _reasonable_ and more performant to combine the parsing & preparation steps and only requires two passes. Since the block is already being traversed for parsing, we could maintain an additional parsing context that takes care of marking elements to be fetched for validation. Besides, this approach seems well-geared towards the libraries that we already use for decoding CBOR (a.k.a [minicbor](https://docs.rs/minicbor/latest/minicbor/)) which provides [a mechanism](https://docs.rs/minicbor/latest/minicbor/fn.decode_with.html) for writing decoders that take an additional mutable context as parameter.
+
+This change would, however, require rewriting many of the decoders in Pallas to support decoding from such a custom context. Since the context wouldn't change the underlying serialization, we could reuse the existing decoders based on the unit (i.e., `()`) context.
```

Comment:
Sounds like digging the best tradeoff solution is a well-identified task that someone could tackle in isolation.

Reply:
Indeed!